File: DSA/Stack/Code/PostfixToInfix.cs
// Postfix to Infix in C#

using System;
using System.Collections.Generic;

class PostfixToInfix {
    static bool IsOperator(char c) {
        return (c == '+' || c == '-' || c == '*' || c == '/');
    }
    
    static string PostfixToInfixConversion(string postfix) {
        Stack<string> stack = new Stack<string>();
        
        foreach (char c in postfix) {
            if (IsOperator(c)) {
                string op2 = stack.Pop();
                string op1 = stack.Pop();
                
                string temp = "(" + op1 + c + op2 + ")";
                stack.Push(temp);
            } else {
                stack.Push(c.ToString());
            }
        }
        
        return stack.Pop();
    }
    
    static void Main() {
        string postfix = "ab+cd*-";
        
        Console.WriteLine("=== Postfix to Infix ===\n");
        Console.WriteLine("Postfix Expression: " + postfix);
        
        string infix = PostfixToInfixConversion(postfix);
        
        Console.WriteLine("Infix Expression: " + infix + "\n");
        Console.WriteLine("Complexity: O(n)");
    }
}

File: DSA/Stack/Code/PostfixToInfix.py
# Postfix to Infix in Python

def is_operator(c):
    return c in ['+', '-', '*', '/']

def postfix_to_infix(postfix):
    stack = []
    
    for c in postfix:
        if is_operator(c):
            op2 = stack.pop()
            op1 = stack.pop()
            
            temp = f"({op1}{c}{op2})"
            stack.append(temp)
        else:
            stack.append(c)
    
    return stack[0]

# Test
postfix = "ab+cd*-"

print("=== Postfix to Infix ===\n")
print(f"Postfix Expression: {postfix}")

infix = postfix_to_infix(postfix)

print(f"Infix Expression: {infix}\n")
print("Complexity: O(n)")
